src/clj/game/cards/agendas.clj

```diff
@@ -140,6 +140,56 @@
                                                                   (clear-wait-prompt :runner))} card nil)))}
                               card nil))}}
 
+   "Bacterial Programming"
+   (letfn [(hq-step [remaining to-trash to-hq]
+             {:delayed-completion true
+              :prompt "Select a card to move to HQ"
+              :choices (conj (vec remaining) "Done")
+              :effect (req (if (= "Done" target)
+                             (do
+                               (doseq [t to-trash]
+                                 (println "Moving" (:title t) "to trash")
+                                 (trash state :corp t {:unpreventable true}))
+                               (doseq [h to-hq]
+                                 (println "Moving" (:title h) "to hand")
+                                 (move state :corp h :hand))
+                               (println "Remaining cards:" (map :title remaining))
+                               (continue-ability state :corp (reorder-choice :corp :runner remaining '()
+                                                                            (count remaining) remaining) card nil)
+                               (system-msg state :corp (str "uses Bacterial Programming to add " (count to-hq) " cards to HQ,  discard "
+                                                            (count to-trash) ", and arrange the top cards of R&D")))
+                             (do
+                               (system-msg state :corp (str "selected " (:title target) " to move to HQ"))
+                               (continue-ability state :corp (hq-step
+                                                               (clojure.set/difference (set remaining) (set [target]))
+                                                               to-trash
+                                                               (conj to-hq target)) card nil))))})
+           (trash-step [remaining to-trash]
+             {:delayed-completion true
+              :prompt "Select a card to discard"
+              :choices (conj (vec remaining) "Done")
+              :effect (req (if (= "Done" target)
+                             (continue-ability state :corp (hq-step remaining to-trash `()) card nil)
+                             (do
+                               (system-msg state :corp (str "selected " (:title target) " to trash"))
+                               (continue-ability state :corp (trash-step
+                                                               (clojure.set/difference (set remaining) (set [target]))
+                                                               (conj to-trash target)) card nil))))})]
+     (let [arrange-rd (effect (continue-ability
+                                {:delayed-completion true
```

Comment:
I think this actually goes inside the `:optional` map, next to the `:prompt`. No keys that are "parallel" to an `:optional` are actually used in the resolution of an optional ability. (See `check-optional` in abilities.clj -- we read the `:optional` key of the given `ability`, then completely ignore anything else that might have also been in `ability`.)